### PR TITLE
Add Stalker.flush().

### PR DESCRIPTION
**Pull request by Ayrx**
This PR adds a Stalker.flush() method that drains the queue.

Ref frida/frida-core#218.

### CODE REVIEW

File: gum/backend-x86/gumstalker-x86.c [hunk 1]
```diff
@@ -569,6 +569,31 @@ gum_stalker_set_trust_threshold (GumStalker * self,
...
+{
+  GSList * sinks, * cur;
+
+  GUM_STALKER_LOCK (self);
+
+  sinks = NULL;
+  for (cur = self->contexts; cur != NULL; cur = cur->next)
+  {
+    GumExecCtx * ctx = cur->data;
+	sinks = g_slist_prepend (sinks, g_object_ref (ctx->sink));
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Should only use spaces for indentation:

```suggestion
    sinks = g_slist_prepend (sinks, g_object_ref (ctx->sink));
```

File: gum/backend-x86/gumstalker-x86.c [hunk 2]
```diff
@@ -569,6 +569,31 @@ gum_stalker_set_trust_threshold (GumStalker * self,
...
+  }
+
+  GUM_STALKER_UNLOCK (self);
+
+  for (cur = sinks; cur != NULL; cur = cur->next)
+  {
+    GumEventSink * sink = cur->data;
+	gum_event_sink_flush (sink);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Should only use spaces for indentation:

```suggestion
    gum_event_sink_flush (sink);
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 3]
```diff
@@ -466,6 +466,31 @@ gum_stalker_set_trust_threshold (GumStalker * self,
...
+{
+  GSList * sinks, * cur;
+
+  GUM_STALKER_LOCK (self);
+
+  sinks = NULL;
+  for (cur = self->contexts; cur != NULL; cur = cur->next)
+  {
+    GumExecCtx * ctx = cur->data;
+	sinks = g_slist_prepend (sinks, g_object_ref (ctx->sink));
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: Should only use spaces for indentation:

```suggestion
    sinks = g_slist_prepend (sinks, g_object_ref (ctx->sink));
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 4]
```diff
@@ -466,6 +466,31 @@ gum_stalker_set_trust_threshold (GumStalker * self,
...
+  }
+
+  GUM_STALKER_UNLOCK (self);
+
+  for (cur = sinks; cur != NULL; cur = cur->next)
+  {
+    GumEventSink * sink = cur->data;
+	gum_event_sink_flush (sink);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: Should only use spaces for indentation:

```suggestion
    gum_event_sink_flush (sink);
```

File: bindings/gumjs/gumdukeventsink.c [hunk 5]
```diff
@@ -173,6 +175,13 @@ gum_duk_event_sink_process (GumEventSink * sink,
   gum_spinlock_release (&self->lock);
 }
 
+static void
+gum_duk_event_sink_flush (GumEventSink * sink)
+{
+  GumDukEventSink * self = GUM_DUK_EVENT_SINK (sink);
```
Comment (oleavr) [inline bindings/gumjs/gumdukeventsink.c]:
```suggestion
```

File: bindings/gumjs/gumdukeventsink.c [hunk 6]
```diff
@@ -173,6 +175,13 @@ gum_duk_event_sink_process (GumEventSink * sink,
...
+  gum_duk_event_sink_drain (self);
```
Comment (oleavr) [inline bindings/gumjs/gumdukeventsink.c]:
```suggestion
  gum_duk_event_sink_drain (GUM_DUK_EVENT_SINK (sink));
```

File: bindings/gumjs/gumv8eventsink.cpp [hunk 7]
```diff
@@ -179,6 +181,13 @@ gum_v8_event_sink_process (GumEventSink * sink,
   gum_spinlock_release (&self->lock);
 }
 
+static void
+gum_v8_event_sink_flush (GumEventSink * sink)
+{
+  auto self = GUM_V8_EVENT_SINK (sink);
```
Comment (oleavr) [inline bindings/gumjs/gumv8eventsink.cpp]:
```suggestion
```

File: bindings/gumjs/gumv8eventsink.cpp [hunk 8]
```diff
@@ -179,6 +181,13 @@ gum_v8_event_sink_process (GumEventSink * sink,
...
+  gum_v8_event_sink_drain (self);
```
Comment (oleavr) [inline bindings/gumjs/gumv8eventsink.cpp]:
```suggestion
  gum_v8_event_sink_drain (GUM_V8_EVENT_SINK (sink));
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 9]
```diff
@@ -466,6 +466,31 @@ gum_stalker_set_trust_threshold (GumStalker * self,
...
+gum_stalker_flush (GumStalker * self)
+{
+  GSList * sinks, * cur;
+
+  GUM_STALKER_LOCK (self);
+
+  sinks = NULL;
+  for (cur = self->contexts; cur != NULL; cur = cur->next)
+  {
+    GumExecCtx * ctx = cur->data;
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
    GumExecCtx * ctx = cur->data;

```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 10]
```diff
@@ -466,6 +466,31 @@ gum_stalker_set_trust_threshold (GumStalker * self,
...
+	sinks = g_slist_prepend (sinks, g_object_ref (ctx->sink));
+  }
+
+  GUM_STALKER_UNLOCK (self);
+
+  for (cur = sinks; cur != NULL; cur = cur->next)
+  {
+    GumEventSink * sink = cur->data;
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
    GumEventSink * sink = cur->data;

```

File: gum/backend-x86/gumstalker-x86.c [hunk 11]
```diff
@@ -569,6 +569,31 @@ gum_stalker_set_trust_threshold (GumStalker * self,
...
+gum_stalker_flush (GumStalker * self)
+{
+  GSList * sinks, * cur;
+
+  GUM_STALKER_LOCK (self);
+
+  sinks = NULL;
+  for (cur = self->contexts; cur != NULL; cur = cur->next)
+  {
+    GumExecCtx * ctx = cur->data;
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]:
```suggestion
    GumExecCtx * ctx = cur->data;

```

File: gum/backend-x86/gumstalker-x86.c [hunk 12]
```diff
@@ -569,6 +569,31 @@ gum_stalker_set_trust_threshold (GumStalker * self,
...
+	sinks = g_slist_prepend (sinks, g_object_ref (ctx->sink));
+  }
+
+  GUM_STALKER_UNLOCK (self);
+
+  for (cur = sinks; cur != NULL; cur = cur->next)
+  {
+    GumEventSink * sink = cur->data;
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]:
```suggestion
    GumEventSink * sink = cur->data;

```

**Comment by Ayrx** [comment]
@oleavr That should be all the changes resolved.

File: gum/gumstalker.h [hunk 13]
```diff
@@ -86,6 +86,7 @@ GUM_API gint gum_stalker_get_trust_threshold (GumStalker * self);
 GUM_API void gum_stalker_set_trust_threshold (GumStalker * self,
     gint trust_threshold);
 
+GUM_API void gum_stalker_flush (GumStalker * self);
```
[the remark this answers is not in the document]
Comment (Ayrx) [reply gum/gumstalker.h]: That should do it.

File: vapi/frida-gum-1.0.vapi [hunk 14]
```diff
@@ -118,6 +118,7 @@ namespace Gum {
 		public int get_trust_threshold ();
 		public void set_trust_threshold (int trust_threshold);
 
+		public void flush();
```
Comment (oleavr) [inline vapi/frida-gum-1.0.vapi]:
```suggestion
		public void flush ();
```

**Comment by oleavr** [comment]
Was about to do a release, so I just [wired it up](https://github.com/frida/frida-core/commit/ca58d8990bb7f4b5fe51901e6bc5aef31a08d5df) in frida-core. Frida 12.2.21 is currently being released with these improvements. Cheers!

**Pull request merged**